jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -2372,7 +2380,93 @@ else if (oldQueryParams == null || oldQueryParams.size() == 0)
     @Override
     public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IOException, ServletException
     {
-        throw new ServletException("HttpServletRequest.upgrade() not supported in Jetty");
+        Response response = _channel.getResponse();
+        if (response.getStatus() != HttpStatus.SWITCHING_PROTOCOLS_101)
+            throw new IllegalStateException("Response status should be 101");
+        if (response.isCommitted())
+            throw new IllegalStateException("Cannot upgrade committed response");
+        if (_metaData == null || _metaData.getHttpVersion() != HttpVersion.HTTP_1_1)
+            throw new IllegalStateException("Only requests over HTTP/1.1 can be upgraded");
```

Comment:
Clients will probably choke if they don't see the `Upgrade` header and `Connection: Upgrade`.
Is it worth to add these checks too?
Also, should we check that there is no content written already?

Reply:
We could add those checks for the `Connection: Upgrade` header, as well as checking that the `Upgrade` header is present, but the servlet developer has to check the contents of the `Upgrade` header before calling `upgrade` to make sure that this protocol can be handled so I don't think they would add much.

Regarding written content, it's again a check we could add but it would prevent mixing a HTTP response followed by a protocol upgrade. Sounds odd but nothing in the servlet nor HTTP specs prevent that.